src/vanilla/store.ts

```diff
@@ -723,43 +708,90 @@ const buildStore = (
+      return [
+        (atom) => {
+          let proxyAtomState = proxyAtomStateMap.get(atom)
+          if (!proxyAtomState) {
+            const atomState = getAtomState(atom)
+            proxyAtomState = new Proxy(atomState, {
+              set(target, prop, value) {
+                if (prop === 'm') {
+                  debugMountedAtoms.add(atom)
+                }
+                return Reflect.set(target, prop, value)
+              },
+              deleteProperty(target, prop) {
+                if (prop === 'm') {
+                  debugMountedAtoms.delete(atom)
+                }
+                return Reflect.deleteProperty(target, prop)
+              },
+            })
+            proxyAtomStateMap.set(atom, proxyAtomState)
           }
-          return atomState
+          return proxyAtomState
         },
-      }),
-      dev4_get_mounted_atoms: () => debugMountedAtoms,
-      dev4_restore_atoms: (values) => {
-        const batch = createBatch()
-        for (const [atom, value] of values) {
-          if (hasInitialValue(atom)) {
-            const atomState = getAtomState(atom)
-            const prevEpochNumber = atomState.n
-            setAtomStateValueOrPromise(atom, atomState, value)
-            mountDependencies(batch, atom, atomState)
-            if (prevEpochNumber !== atomState.n) {
-              registerBatchAtom(batch, atom, atomState)
-              recomputeDependents(batch, atom, atomState)
-            }
+        atomRead,
+        (atom, getter, setter, ...args) => {
+          if (inRestoreAtom) {
+            return setter(atom, ...args)
           }
+          return atomWrite(atom, getter, setter, ...args)
+        },
+        atomOnMount,
+      ]
+    },
+  )
+  const savedStoreSet = derivedStore.set
+  const devStore: DevStoreRev4 = {
+    // store dev methods (these are tentative and subject to change without notice)
+    dev4_get_internal_weak_map: () => ({
+      get: (atom) => {
+        const atomState = savedGetAtomState(atom)
+        if (atomState.n === 0) {
+          // for backward compatibility
+          return undefined
         }
-        flushBatch(batch)
+        return atomState
       },
-    }
-    Object.assign(store, devStore)
+    }),
+    dev4_get_mounted_atoms: () => debugMountedAtoms,
+    dev4_restore_atoms: (values) => {
+      const restoreAtom: WritableAtom<null, [], void> = {
+        read: () => null,
+        write: (_get, set) => {
+          ++inRestoreAtom
+          try {
+            for (const [atom, value] of values) {
+              if (hasInitialValue(atom)) {
+                set(atom as never, value)
+              }
+            }
+          } finally {
+            --inRestoreAtom
+          }
+        },
+      }
+      savedStoreSet(restoreAtom)
+    },
   }
-  return store
+  return Object.assign(derivedStore, devStore)
 }
 
-export const createStore = (): Store => {
+type PrdOrDevStore = Store | (Store & DevStoreRev4)
```

Comment:
I think type `Store` should represent the complete store union. It makes the type cleaner in user land.
<img width="583" alt="image" src="https://github.com/user-attachments/assets/49b6ca7e-2ef3-479a-acf0-dad83c007a5f" />

```suggestion
type Store = PrdStore | (PrdStore & DevStoreRev4)
```

Reply:
It just looks ugly, right?
I'm planning to change the `createStore` to return just store without dev methods, in the near future (kind of breaking change to `jotai-devtools`). So, this is somewhat intentional.

Reply:
> It just looks ugly, right?

Yes

> I'm planning to change the `createStore` to return just store without dev methods,

I just think the type returned from `createStore` should be called `Store`.

Reply:
> I just think the type returned from `createStore` should be called `Store`.

So, it will in the near future. Meanwhile, this ugly type is kind of warning. I don't know if anybody realize it though.